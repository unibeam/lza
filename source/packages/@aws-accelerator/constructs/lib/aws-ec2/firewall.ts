/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { LaunchTemplateConfig, NetworkInterfaceItemConfig } from '@aws-accelerator/config';
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import fs from 'fs';
import path from 'path';
import { LaunchTemplate } from './create-launch-template';

export interface IFirewall extends cdk.IResource {
  /**
   * The friendly name of the firewall instance
   */
  readonly name: string;
}

export interface FirewallProps {
  /**
   * The friendly name of the firewall instance
   */
  readonly name: string;
  /**
   * The name of the firewall configuration bucket
   */
  readonly configBucketName: string;
  /**
   * The configuration directory path
   */
  readonly configDir: string;
  /**
   * The launch template for the firewall instance
   */
  readonly launchTemplate: LaunchTemplateConfig;
  /**
   * The name of the VPC to deploy the instance to
   */
  readonly vpc: string;
  /**
   * An array of CloudFormation tags
   */
  readonly tags?: cdk.CfnTag[];
}

export class Firewall extends cdk.Resource implements IFirewall {
  public readonly name: string;
  protected publicIpAddresses: Map<number, string> = new Map<number, string>();
  protected launchTemplate: LaunchTemplate;
  protected networkInterfaces: NetworkInterfaceItemConfig[];
  protected props: FirewallProps;
  private eipAssociations: cdk.aws_ec2.CfnEIPAssociation[] = [];

  constructor(scope: Construct, id: string, props: FirewallProps) {
    super(scope, id);
    this.name = props.name;
    this.props = props;

    // Create interfaces with elastic IPs, if needed
    this.networkInterfaces = this.setNetworkInterfaceProps();
    // Create launch template
    this.launchTemplate = this.createLaunchTemplate();
    // Create EIP interface dependencies
    this.eipAssociations.forEach(association => this.launchTemplate.node.addDependency(association));
  }

  /**
   * Create a launch template for the firewall
   * @returns
   */
  private createLaunchTemplate(): LaunchTemplate {
    return new LaunchTemplate(this, 'LaunchTemplate', {
      appName: this.props.name,
      name: this.props.launchTemplate.name,
      vpc: this.props.vpc,
      blockDeviceMappings: this.props.launchTemplate.blockDeviceMappings,
      userData: this.processUserData(this.props.launchTemplate.userData),
      securityGroups: this.props.launchTemplate.securityGroups,
      networkInterfaces: this.networkInterfaces,
      instanceType: this.props.launchTemplate.instanceType,
      keyPair: this.props.launchTemplate.keyPair,
      iamInstanceProfile: this.props.launchTemplate.iamInstanceProfile,
      imageId: this.props.launchTemplate.imageId,
      enforceImdsv2: this.props.launchTemplate.enforceImdsv2,
    });
  }

  /**
   * Set network interface properties for the launch template
   * @returns
   */
  private setNetworkInterfaceProps(): NetworkInterfaceItemConfig[] {
    const networkInterfaces: NetworkInterfaceItemConfig[] = [];
    let deviceIndex = 0;
    for (const networkInterface of this.props.launchTemplate.networkInterfaces ?? []) {
      // Create interface with elastic IP
      if (networkInterface.associateElasticIp) {
        networkInterfaces.push(
          this.createEipInterface(
            networkInterface,
            networkInterface.deviceIndex !== undefined ? networkInterface.deviceIndex : deviceIndex,
          ),
        );
        deviceIndex += 1;
        continue;
      }

      // Create interface with source/dest check disabled
      if (!networkInterface.associateElasticIp && networkInterface.sourceDestCheck === false) {
        networkInterfaces.push(
          this.createRouterInterface(
            networkInterface,
            networkInterface.deviceIndex !== undefined ? networkInterface.deviceIndex : deviceIndex,
          ),
        );
        deviceIndex += 1;
        continue;
      }

      networkInterfaces.push({
        associateCarrierIpAddress: networkInterface.associateCarrierIpAddress,
        associateElasticIp: undefined,
        associatePublicIpAddress: networkInterface.associatePublicIpAddress,
        deleteOnTermination: networkInterface.deleteOnTermination,
        description: networkInterface.description,
        deviceIndex: networkInterface.deviceIndex,
        groups: networkInterface.groups,
        interfaceType: networkInterface.interfaceType,
        networkCardIndex: networkInterface.networkCardIndex,
        networkInterfaceId: networkInterface.networkInterfaceId,
        privateIpAddress: networkInterface.privateIpAddress,
        privateIpAddresses: networkInterface.privateIpAddresses,
        secondaryPrivateIpAddressCount: networkInterface.secondaryPrivateIpAddressCount,
        sourceDestCheck: undefined,
        subnetId: networkInterface.subnetId,
      });
      deviceIndex += 1;
    }
    return networkInterfaces;
  }

  /**
   * Create and associate an EIP with a network interface
   * @param networkInterface
   * @param deviceIndex
   * @returns
   */
  private createEipInterface(
    networkInterface: NetworkInterfaceItemConfig,
    deviceIndex: number,
  ): NetworkInterfaceItemConfig {
    // Create EIP
    const eip = new cdk.aws_ec2.CfnEIP(this, `ElasticIp${deviceIndex}`, {
      domain: 'vpc',
    });

    // Add public IP to map
    this.publicIpAddresses.set(deviceIndex, eip.ref);

    // Create interface
    const eipInterface = new cdk.aws_ec2.CfnNetworkInterface(this, `NetworkInterface${deviceIndex}`, {
      description: networkInterface.description,
      groupSet: networkInterface.groups,
      interfaceType: networkInterface.interfaceType,
      privateIpAddress: networkInterface.privateIpAddress,
      privateIpAddresses: networkInterface.privateIpAddresses
        ? (networkInterface.privateIpAddresses as cdk.aws_ec2.CfnNetworkInterface.PrivateIpAddressSpecificationProperty[])
        : undefined,
      secondaryPrivateIpAddressCount: networkInterface.secondaryPrivateIpAddressCount,
      sourceDestCheck: networkInterface.sourceDestCheck,
      subnetId: networkInterface.subnetId!,
    });

    // Associate EIP
    this.eipAssociations.push(
      new cdk.aws_ec2.CfnEIPAssociation(this, `EipAssociation${deviceIndex}`, {
        allocationId: eip.attrAllocationId,
        networkInterfaceId: eipInterface.ref,
      }),
    );

    return {
      deviceIndex: deviceIndex,
      networkInterfaceId: eipInterface.ref,
    } as NetworkInterfaceItemConfig;
  }

  /**
   * Create a network interface with source/dest checks disabled
   * @param networkInterface
   * @param deviceIndex
   * @returns
   */
  private createRouterInterface(
    networkInterface: NetworkInterfaceItemConfig,
    deviceIndex: number,
  ): NetworkInterfaceItemConfig {
    // Create interface
    const routerInterface = new cdk.aws_ec2.CfnNetworkInterface(this, `NetworkInterface${deviceIndex}`, {
      description: networkInterface.description,
      groupSet: networkInterface.groups,
      interfaceType: networkInterface.interfaceType,
      privateIpAddress: networkInterface.privateIpAddress,
      privateIpAddresses: networkInterface.privateIpAddresses
        ? (networkInterface.privateIpAddresses as cdk.aws_ec2.CfnNetworkInterface.PrivateIpAddressSpecificationProperty[])
        : undefined,
      secondaryPrivateIpAddressCount: networkInterface.secondaryPrivateIpAddressCount,
      sourceDestCheck: networkInterface.sourceDestCheck,
      subnetId: networkInterface.subnetId!,
    });

    return {
      deviceIndex: deviceIndex,
      networkInterfaceId: routerInterface.ref,
    } as NetworkInterfaceItemConfig;
  }

  /**
   * Returns a base-64 encoded userdata string, replacing the
   * firewall config bucket name, if applicable
   * @param userdataPath string | undefined
   * @returns string | undefined
   */
  private processUserData(userdataPath?: string): string | undefined {
    if (!userdataPath) {
      return;
    }
    // Replace config bucket name variable
    const varRegex = new RegExp('\\$\\{ACCEL_LOOKUP::S3:BUCKET:firewall-config\\}', 'gi');
    const userdata = fs.readFileSync(path.join(this.props.configDir, userdataPath), 'utf-8');

    return cdk.Fn.base64(userdata.replace(varRegex, this.props.configBucketName));
  }
}
