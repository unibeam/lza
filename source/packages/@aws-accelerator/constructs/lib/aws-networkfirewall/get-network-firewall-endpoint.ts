/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

interface IGetNetworkFirewallEndpoint extends cdk.IResource {
  /**
   * The ID of the endpoint
   */
  readonly endpointId: string;
}

interface GetNetworkFirewallEndpointProps {
  /**
   * The AZ the endpoint is located in
   */
  readonly endpointAz: string;

  /**
   * The ARN of the associated Network Firewall
   */
  readonly firewallArn: string;

  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;

  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;

  /**
   * The region of the Network Firewall
   */
  readonly region: string;
}

export class GetNetworkFirewallEndpoint extends cdk.Resource implements IGetNetworkFirewallEndpoint {
  public readonly endpointId: string;

  constructor(scope: Construct, id: string, props: GetNetworkFirewallEndpointProps) {
    super(scope, id);

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, 'Custom::GetNetworkFirewallEndpoint', {
      codeDirectory: path.join(__dirname, 'get-network-firewall-endpoint/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Effect: 'Allow',
          Action: ['ec2:DescribeAvailabilityZones', 'network-firewall:DescribeFirewall'],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: 'Custom::GetNetworkFirewallEndpoint',
      serviceToken: provider.serviceToken,
      properties: {
        endpointAz: props.endpointAz,
        firewallArn: props.firewallArn,
        region: props.region,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.endpointId = resource.ref;
  }
}
