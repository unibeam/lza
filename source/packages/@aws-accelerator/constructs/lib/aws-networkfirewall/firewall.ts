/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { GetNetworkFirewallEndpoint } from './get-network-firewall-endpoint';

export interface INetworkFirewall extends cdk.IResource {
  /**
   * The unique IDs of the firewall endpoints for all of the subnets that you attached to the firewall.
   */
  readonly endpointIds?: string[];
  /**
   * The Amazon Resource Name (ARN) of the firewall.
   */
  readonly firewallArn: string;

  /**
   * The ID of the firewall.
   */
  readonly firewallId?: string;

  /**
   * The name of the policy.
   */
  readonly firewallName: string;

  addLogging: (config: cdk.aws_networkfirewall.CfnLoggingConfiguration.LoggingConfigurationProperty) => void;

  addNetworkFirewallRoute: (
    id: string,
    endpointAz: string,
    logRetentionInDays: number,
    routeTableId: string,
    destination?: string,
    ipv6Destination?: string,
    logGroupKmsKey?: cdk.aws_kms.IKey,
  ) => void;
}

interface NetworkFirewallProps {
  /**
   * The Amazon Resource Name (ARN) of the firewall policy.
   */
  readonly firewallPolicyArn: string;

  /**
   * The descriptive name of the firewall.
   */
  readonly name: string;

  /**
   * The subnets that Network Firewall is using for the firewall.
   */
  readonly subnets: string[];

  /**
   * The unique identifier of the VPC where the firewall is in use.
   */
  readonly vpcId: string;

  /**
   * A flag indicating whether it is possible to delete the firewall.
   */
  readonly deleteProtection?: boolean;

  /**
   * A description of the firewall.
   */
  readonly description?: string;

  /**
   * A setting indicating whether the firewall is protected against a change to the firewall policy association.
   */
  readonly firewallPolicyChangeProtection?: boolean;

  /**
   * A setting indicating whether the firewall is protected against changes to the subnet associations.
   */
  readonly subnetChangeProtection?: boolean;

  /**
   * An optional list of CloudFormation tags.
   */
  readonly tags?: cdk.CfnTag[];
}

abstract class NetworkFirewallBase extends cdk.Resource implements INetworkFirewall {
  public abstract readonly firewallArn: string;
  public abstract readonly firewallName: string;
  public abstract readonly firewallId?: string;
  public abstract readonly endpointIds?: string[];
  // protected abstract subnetMapping?: cdk.aws_networkfirewall.CfnFirewall.SubnetMappingProperty[];

  public addLogging(config: cdk.aws_networkfirewall.CfnLoggingConfiguration.LoggingConfigurationProperty) {
    new cdk.aws_networkfirewall.CfnLoggingConfiguration(this, 'LoggingConfig', {
      firewallArn: this.firewallArn,
      loggingConfiguration: config,
    });
  }

  public addNetworkFirewallRoute(
    id: string,
    endpointAz: string,
    logRetentionInDays: number,
    routeTableId: string,
    destination?: string,
    ipv6Destination?: string,
    logGroupKmsKey?: cdk.aws_kms.IKey,
  ): void {
    // Get endpoint ID from custom resource
    const vpcEndpointId = new GetNetworkFirewallEndpoint(this, `${id}Endpoint`, {
      endpointAz: endpointAz,
      firewallArn: this.firewallArn,
      kmsKey: logGroupKmsKey,
      logRetentionInDays: logRetentionInDays,
      region: cdk.Stack.of(this).region,
    }).endpointId;

    new cdk.aws_ec2.CfnRoute(this, id, {
      routeTableId,
      destinationCidrBlock: destination,
      destinationIpv6CidrBlock: ipv6Destination,
      vpcEndpointId,
    });
  }
}
export class NetworkFirewall extends NetworkFirewallBase {
  public readonly firewallArn: string;
  public readonly firewallName: string;
  public readonly firewallId?: string;
  public readonly endpointIds?: string[];
  protected subnetMapping: cdk.aws_networkfirewall.CfnFirewall.SubnetMappingProperty[];

  static fromAttributes(
    scope: Construct,
    id: string,
    attrs: {
      firewallArn: string;
      firewallName: string;
    },
  ): INetworkFirewall {
    class Import extends NetworkFirewallBase {
      public readonly firewallArn: string;
      public readonly firewallName: string;
      public readonly firewallId?: string;
      public readonly endpointIds?: string[];
      constructor(scope: Construct, id: string) {
        super(scope, id);
        this.firewallArn = attrs.firewallArn;
        this.firewallName = attrs.firewallName;
      }
    }
    return new Import(scope, id);
  }

  /**
   * Returns CfnFirewall by applying updates to included resource
   * @param scope Stack in which included Firewall is created/managed
   * @param id logicalId of Firewall
   * @param attrs
   */
  static includedCfnResource(scope: cdk.cloudformation_include.CfnInclude, id: string, props: NetworkFirewallProps) {
    // Set initial properties
    const subnetMapping = props.subnets.map(item => {
      return { subnetId: item };
    });
    const resource = scope.getResource(id) as cdk.aws_networkfirewall.CfnFirewall;
    resource.deleteProtection = props.deleteProtection;
    resource.subnetChangeProtection = props.subnetChangeProtection;
    resource.subnetMappings = subnetMapping;
    resource.firewallPolicyArn = props.firewallPolicyArn;
    resource.firewallPolicyChangeProtection = props.firewallPolicyChangeProtection;
    resource.description = props.description;
    return resource;
  }

  constructor(scope: Construct, id: string, props: NetworkFirewallProps) {
    super(scope, id);

    // Set initial properties
    this.firewallName = props.name;
    this.subnetMapping = props.subnets.map(item => {
      return { subnetId: item };
    });

    // Set name tag
    props.tags?.push({ key: 'Name', value: this.firewallName });

    const resource = new cdk.aws_networkfirewall.CfnFirewall(this, 'Resource', {
      firewallName: this.firewallName,
      firewallPolicyArn: props.firewallPolicyArn,
      subnetMappings: this.subnetMapping,
      vpcId: props.vpcId,
      deleteProtection: props.deleteProtection,
      description: props.description,
      firewallPolicyChangeProtection: props.firewallPolicyChangeProtection,
      subnetChangeProtection: props.subnetChangeProtection,
      tags: props.tags,
    });

    // Set remaining properties
    this.endpointIds = resource.attrEndpointIds;
    this.firewallArn = resource.ref;
    this.firewallId = resource.attrFirewallId;
  }
}
