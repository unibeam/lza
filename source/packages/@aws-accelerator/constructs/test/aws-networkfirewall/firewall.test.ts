/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import { NetworkFirewall } from '../../lib/aws-networkfirewall/firewall';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(NetworkFirewall): ';

//Initialize stack for resource configuration test
const stack = new cdk.Stack();

const firewallPolicyArn = 'arn:aws:network-firewall:us-east-1:222222222222:firewall-policy/TestPolicy';

const importedFirewallArn = 'arn:aws:network-firewall:us-east-1:222222222222:firewall/TestImportedFirewall';

const testFirewall = new NetworkFirewall(stack, 'TestFirewall', {
  firewallPolicyArn: firewallPolicyArn,
  name: 'TestFirewall',
  subnets: ['Test-Subnet-1', 'Test-Subnet-2'],
  vpcId: 'TestVpc',
  tags: [],
});
const key = new cdk.aws_kms.Key(stack, 'CloudWatchKey', {});

testFirewall.addNetworkFirewallRoute('testRouteId1', '1', 365, 'routeTableId', '10.0.0.6/32', undefined, key);
testFirewall.addNetworkFirewallRoute('testRouteId2', '1', 365, 'routeTableId', undefined, '::1', key);

const importedFirewall = NetworkFirewall.fromAttributes(stack, 'TestImportFirewall', {
  firewallArn: importedFirewallArn,
  firewallName: 'ImportedFirewallName',
});
importedFirewall.addLogging({
  logDestinationConfigs: [
    {
      logDestinationType: 'CloudWatchLogs',
      logDestination: {
        logGroup: 'firewallAlertLogGroupArn',
      },
      logType: 'ALERT',
    },
  ],
});

importedFirewall.addNetworkFirewallRoute('endpointRouteId', '1', 365, 'routeTableId', '10.0.0.6/32', undefined, key);

const app = new cdk.App();
const includedStack = new cdk.Stack(app, `placeHolder`, {});
const firewallStack = new cdk.cloudformation_include.CfnInclude(includedStack, 'IncludedStack', {
  templateFile: path.join(__dirname, 'includedStacks/firewall-stack.json'),
});

NetworkFirewall.includedCfnResource(firewallStack, 'firewallLogicalId', {
  firewallPolicyArn: firewallPolicyArn,
  name: 'TestFirewall',
  subnets: ['Test-Subnet-1', 'Test-Subnet-2'],
  vpcId: 'TestVpc',
  tags: [],
});

/**
 * Network Firewall construct test
 */
describe('Network Firewall', () => {
  snapShotTest(testNamePrefix, stack);
  snapShotTest(testNamePrefix, includedStack);
});
