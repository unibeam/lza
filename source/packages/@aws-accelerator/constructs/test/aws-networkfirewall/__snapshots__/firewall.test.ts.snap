// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network Firewall Construct(NetworkFirewall):  Snapshot Test 1`] = `
{
  "Resources": {
    "CloudWatchKey9B181885": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomGetNetworkFirewallEndpointCustomResourceProviderHandler2EF030A1": {
      "DependsOn": [
        "CustomGetNetworkFirewallEndpointCustomResourceProviderRole540B9917",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomGetNetworkFirewallEndpointCustomResourceProviderRole540B9917",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGetNetworkFirewallEndpointCustomResourceProviderLogGroup98AC3B14": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CloudWatchKey9B181885",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomGetNetworkFirewallEndpointCustomResourceProviderHandler2EF030A1",
              },
            ],
          ],
        },
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomGetNetworkFirewallEndpointCustomResourceProviderRole540B9917": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ec2:DescribeAvailabilityZones",
                    "network-firewall:DescribeFirewall",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestFirewallE26FCA5C": {
      "Properties": {
        "FirewallName": "TestFirewall",
        "FirewallPolicyArn": "arn:aws:network-firewall:us-east-1:222222222222:firewall-policy/TestPolicy",
        "SubnetMappings": [
          {
            "SubnetId": "Test-Subnet-1",
          },
          {
            "SubnetId": "Test-Subnet-2",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestFirewall",
          },
        ],
        "VpcId": "TestVpc",
      },
      "Type": "AWS::NetworkFirewall::Firewall",
    },
    "TestFirewalltestRouteId1894E055E": {
      "Properties": {
        "DestinationCidrBlock": "10.0.0.6/32",
        "RouteTableId": "routeTableId",
        "VpcEndpointId": {
          "Ref": "TestFirewalltestRouteId1Endpoint4D98AAD2",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestFirewalltestRouteId1Endpoint4D98AAD2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomGetNetworkFirewallEndpointCustomResourceProviderLogGroup98AC3B14",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomGetNetworkFirewallEndpointCustomResourceProviderHandler2EF030A1",
            "Arn",
          ],
        },
        "endpointAz": "1",
        "firewallArn": {
          "Ref": "TestFirewallE26FCA5C",
        },
        "region": {
          "Ref": "AWS::Region",
        },
      },
      "Type": "Custom::GetNetworkFirewallEndpoint",
      "UpdateReplacePolicy": "Delete",
    },
    "TestFirewalltestRouteId280883F56": {
      "Properties": {
        "DestinationIpv6CidrBlock": "::1",
        "RouteTableId": "routeTableId",
        "VpcEndpointId": {
          "Ref": "TestFirewalltestRouteId2Endpoint9D8D500E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestFirewalltestRouteId2Endpoint9D8D500E": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomGetNetworkFirewallEndpointCustomResourceProviderLogGroup98AC3B14",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomGetNetworkFirewallEndpointCustomResourceProviderHandler2EF030A1",
            "Arn",
          ],
        },
        "endpointAz": "1",
        "firewallArn": {
          "Ref": "TestFirewallE26FCA5C",
        },
        "region": {
          "Ref": "AWS::Region",
        },
      },
      "Type": "Custom::GetNetworkFirewallEndpoint",
      "UpdateReplacePolicy": "Delete",
    },
    "TestImportFirewallLoggingConfigB02138AE": {
      "Properties": {
        "FirewallArn": "arn:aws:network-firewall:us-east-1:222222222222:firewall/TestImportedFirewall",
        "LoggingConfiguration": {
          "LogDestinationConfigs": [
            {
              "LogDestination": {
                "logGroup": "firewallAlertLogGroupArn",
              },
              "LogDestinationType": "CloudWatchLogs",
              "LogType": "ALERT",
            },
          ],
        },
      },
      "Type": "AWS::NetworkFirewall::LoggingConfiguration",
    },
    "TestImportFirewallendpointRouteId4582BF89": {
      "Properties": {
        "DestinationCidrBlock": "10.0.0.6/32",
        "RouteTableId": "routeTableId",
        "VpcEndpointId": {
          "Ref": "TestImportFirewallendpointRouteIdEndpoint3D0FC334",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "TestImportFirewallendpointRouteIdEndpoint3D0FC334": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomGetNetworkFirewallEndpointCustomResourceProviderLogGroup98AC3B14",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomGetNetworkFirewallEndpointCustomResourceProviderHandler2EF030A1",
            "Arn",
          ],
        },
        "endpointAz": "1",
        "firewallArn": "arn:aws:network-firewall:us-east-1:222222222222:firewall/TestImportedFirewall",
        "region": {
          "Ref": "AWS::Region",
        },
      },
      "Type": "Custom::GetNetworkFirewallEndpoint",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`Network Firewall Construct(NetworkFirewall):  Snapshot Test 2`] = `
{
  "Resources": {
    "firewallLogicalId": {
      "Properties": {
        "FirewallName": "TestFirewall",
        "FirewallPolicyArn": "arn:aws:network-firewall:us-east-1:222222222222:firewall-policy/TestPolicy",
        "SubnetMappings": [
          {
            "SubnetId": "Test-Subnet-1",
          },
          {
            "SubnetId": "Test-Subnet-2",
          },
        ],
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestFirewall",
          },
        ],
        "VpcId": "TestVpc",
      },
      "Type": "AWS::NetworkFirewall::Firewall",
    },
  },
}
`;
